react/features/chat/middleware.js

```diff
@@ -116,44 +152,183 @@ function _addChatMsgListener(conference, { dispatch, getState }) {
-
-            dispatch(addMessage({
-                displayName,
-                hasRead,
+    conference.on(
+        JitsiConferenceEvents.PRIVATE_MESSAGE_RECEIVED,
+        (id, message, timestamp) => {
+            _handleReceivedMessage({
+                dispatch,
+                getState
+            }, {
                 id,
-                messageType: participant.local ? 'local' : 'remote',
                 message,
-                timestamp: millisecondsTimestamp
-            }));
+                privateMessage: true,
+                timestamp,
+                nick: undefined
+            });
+        }
+    );
+}
+
+/**
+ * Function to handle an incoming chat message.
+ *
+ * @param {Store} store - The Redux store.
+ * @param {Object} message - The message object.
+ * @returns {void}
+ */
+function _handleReceivedMessage({ dispatch, getState }, { id, message, nick, privateMessage, timestamp }) {
+    // Logic for all platforms:
+    const state = getState();
+    const { isOpen: isChatOpen } = state['features/chat'];
 
-            if (typeof APP !== 'undefined') {
-                // Logic for web only:
+    if (!isChatOpen) {
+        dispatch(playSound(INCOMING_MSG_SOUND_ID));
+    }
 
-                APP.API.notifyReceivedChatMessage({
-                    body: message,
-                    id,
-                    nick: displayName,
-                    ts: timestamp
-                });
+    // Provide a default for for the case when a message is being
+    // backfilled for a participant that has left the conference.
+    const participant = getParticipantById(state, id) || {};
+    const localParticipant = getLocalParticipant(getState);
+    const displayName = participant.name || nick || getParticipantDisplayName(state, id);
+    const hasRead = participant.local || isChatOpen;
+    const timestampToDate = timestamp
+        ? new Date(timestamp) : new Date();
+    const millisecondsTimestamp = timestampToDate.getTime();
 
-                dispatch(showToolbox(4000));
-            }
-        }
-    );
+    dispatch(addMessage({
+        displayName,
+        hasRead,
+        id,
+        messageType: participant.local ? 'local' : 'remote',
+        message,
+        privateMessage,
+        recipient: getParticipantDisplayName(state, localParticipant.id),
+        timestamp: millisecondsTimestamp
+    }));
+
+    if (typeof APP !== 'undefined') {
+        // Logic for web only:
+
+        APP.API.notifyReceivedChatMessage({
+            body: message,
+            id,
+            nick: displayName,
+            ts: timestamp
+        });
+
+        dispatch(showToolbox(4000));
+    }
+}
+
+/**
+ * Focuses the chat text field on web after the message recipient was updated, if needed.
+ *
+ * @returns {void}
+ */
+function _maybeFocusField() {
+    if (navigator.product !== 'ReactNative') {
+        const textField = document.getElementById('usermsg');
+
+        textField && textField.focus();
```

Comment:
Could/should this be handled by the text input component itself or maybe one of its parents? It could listen for changes to the private message target and manually focus.

Reply:
![image](https://user-images.githubusercontent.com/9974975/66395648-45840b00-e9d8-11e9-8319-771daf8e4c96.png)

Reply:
So you're not going to address this global DOM accessing?